pkg/util/cryptsetup.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2019 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+	"strings"
+)
+
+// LuksFormat sets up volume as an encrypted LUKS partition
+func LuksFormat(devicePath, passphrase string) (stdout, stderr []byte, err error) {
+	return execCryptsetupCommand(&passphrase, "-q", "luksFormat", "--hash", "sha256", devicePath, "-d", "/dev/stdin")
+}
+
+// LuksOpen opens LUKS encrypted partition and sets up a mapping
+func LuksOpen(devicePath, mapperFile, passphrase string) (stdout, stderr []byte, err error) {
+	return execCryptsetupCommand(&passphrase, "luksOpen", devicePath, mapperFile, "-d", "/dev/stdin")
+}
+
+// LuksClose removes existing mapping
+func LuksClose(mapperFile string) (stdout, stderr []byte, err error) {
+	return execCryptsetupCommand(nil, "luksClose", mapperFile)
```

Comment:
@vasyl-purchel suppose the luksClose fails for some reason ( errors like 'still mapped' or 'resource busy'..etc) , do we need to do forceful close ?

Reply:
@vasyl-purchel also what happens to the mapperfile after close. Is the "/dev" entry for that remains in the system as stale?

Reply:
as far as I know if luksClose failed because of resource is busy there is no real "force" close available. I believe it might be possible to automate some steps but this is not an easy task and usually requires human decision - you need to find what keeps the resource busy, free the resource and after you should be able to do luksClose with no issues

regarding mapperfiles - from what I've seen /dev/mapper/... devices are removed, and then we handle /dev/rbd... device as if there was no such thing as mapper device, so it's detached as if there was no encryption.

Reply:
Ok., as long as you confirm, there is no stale entries remains in the system in a close/failed operation, I am fine.

Reply:
if close operation fails, the mapper device will remain and the /dev/rbd-... device won't be able to be removed as it's still busy too. it will remain as an issue until manual intervention to successfully close it...